### PR TITLE
powershell.ps1: Ensure Fail-Json() works with Hashtables

### PR DESCRIPTION
**Pull request by dagwieers**
##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

powershell.ps1

##### ANSIBLE VERSION

v2.3

##### SUMMARY

Without this change a dictionary $result object would be emptied if it
is anything but a PSObject. Now we also support Hashtables without
losing $result information when we fail with Fail-Json.

### CODE REVIEW

File: lib/ansible/module_utils/powershell.ps1 [hunk 1]
```diff
@@ -67,23 +67,26 @@ Function Exit-Json($obj)
...
 Function Fail-Json($obj, $message = $null)
 {
-    # If we weren't given 2 args, and the only arg was a string, create a new
-    # psobject and use the arg as the failure message
-    If ($message -eq $null -and $obj.GetType().Name -eq "String")
-    {
+    If (-not $obj -or -not $obj.GetType) {
+        # If the first argument is undefined or not an object, make it a Hashtable
+        $obj = @{ }
+    } ElseIf ($obj.GetType().Name -eq "PSCustomObject" -or $obj.GetType().Name -eq "Hashtable") {
```
Comment (jhawkesworth) [inline lib/ansible/module_utils/powershell.ps1]: Would be nice to use the -is test here 
so
`$obj -is [psobject] -or $obj -is [hashtable]`
instead of the `.GetType().Name -eq "name of type as a string"` syntax - in fact that ought to be another thing for ansible powershell style guide IMO.

**Comment by jhawkesworth** [comment]
Thanks for finding this.  
Please see in-line comment - @nitzmahone pointed out that the .GetType().Name -eq "name of type as string" syntax was a bit unnecessary a while back.  Might be better to remove all occurrences in a separate PR, thinking about it.

File: lib/ansible/module_utils/powershell.ps1 (hunk 1 above)
Comment (dagwieers) [reply lib/ansible/module_utils/powershell.ps1]: I didn't even know that was possible :-) All my searches guided me to GetType().
Will fix !

**Comment by dagwieers** [comment]
ready_for_review

**Comment by jhawkesworth** [comment]
Thanks - the powershell.ps1 changes look good now, and I'm keen to get this in.

The change from
`    - asyncresult.changed == false		` 
to
`    - asyncresult.changed == true`

concerns me a little, so I want to try this out before up voting.

**Comment by dagwieers** [comment]
@jhawkesworth The reason why I had to change this, is because previously modules using hashtables would Fail with an empty $result (rather than retain the information in $result). We fixed this now, which means that suddenly all modules using a Hashtable will have different (correct) output from the module when failing. In this case there was a change registered, but we never saw this information because of the initialised $result.

Thanks to fixing this (for win_feature where this behaviour also triggered an integration test to fail) we found only one other integration test that would trip on this. It's not that abnormal that so few tests notice a change, because it only affects the output when failing (and often we don't test if something changed, or any other output when failing).

**Comment by jhawkesworth** [comment]
Thanks for the explanation. Its a good point that we could be checking more on the failure paths in the integration tests.

**Comment by dagwieers** [comment]
@jhawkesworth Definitely, my next job is to add integration tests for modules that have none, and also look what is the best way to do idempotency and check-mode testing. It would be nice if we didn't had to write specific tests for each. Maybe the testing working group already discussed this before.
